### PR TITLE
[FSDP] Implement apply()

### PR DESCRIPTION
**Pull request by rohan-varma**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #72600

Implements `apply()` which applies a `callable` of signature `f(m: Module) -> None` recursively to every submodule. The main difference from `nn.module.apply` is that this version summons the full parameters before apply() so it works appropriately with FSDP.

Closes https://github.com/pytorch/pytorch/issues/72554

Differential Revision: [D34111109](https://our.internmc.facebook.com/intern/diff/D34111109/)

### CODE REVIEW

File: torch/distributed/fsdp/fully_sharded_data_parallel.py [hunk 1]
```diff
@@ -287,6 +287,45 @@ def module(self) -> FlattenParamsWrapper:
...
+
+        Args:
+            fn (:class:`Module` -> None): function to be applied to each submodule
+
+        Returns:
+            Module: self
+        """
+        uninitialized = self._is_root is None
+        self._assert_state(TrainingState_.IDLE)
+        with self._summon_full_params(recurse=False):
```
Comment (zhaojuanmao) [inline torch/distributed/fsdp/fully_sharded_data_parallel.py:318]: hmm, the comment said 'Applies ``fn`` recursively to every submodule', but here _summon_full_params is called with recurse=False, not sure how it works

**Review by zhaojuanmao — APPROVED** [review]
looks good, just having one question

File: torch/distributed/fsdp/fully_sharded_data_parallel.py (hunk 1 above)
Comment (rohan-varma) [reply torch/distributed/fsdp/fully_sharded_data_parallel.py:318]: It is similar to how `state_dict` will be implemented, we don't recurse as we call `super().apply()` which calls the method in nn.Module. That method should run `fn` on this module and then recursively call in on its children modules, so as a result the FSDP instances will have `apply` invoked so it makes sense for it to be nonrecursive.

Comment (zhaojuanmao) [reply torch/distributed/fsdp/fully_sharded_data_parallel.py:318]: I see, that makes sense! Thanks for explaining

**Pull request closed without merging**